### ISSUE TITLE
Difficulties vertically aligning inline latexmath

### ISSUE THREAD

**Issue by davidplowman** [asciidoctor/asciidoctor-pdf#1483]
Hi, I was advised in the asciidoctor-mathematical issues list to post this report here.

I have made a very simple example, consisting of a file with a single line.

`An inline equation latexmath:[w] and bigger latexmath:[{\large w}]. With subscripts latexmath:[w_{ij}] bigger latexmath:[{\large w_{ij}}]. (As plain text, not an equation, we get _w~ij~_)`

and turn it into a pdf using

asciidoctor-pdf -r asciidoctor-mathematical -o test.pdf test.adoc

My problem is that the inline equations (whether using png or svg mode) don't sit on the line with the other text - and it gets worse when there are subscripts. Does anyone know of any techniques to work around this situation?

Here's a screenshoot of how it looks:
![screenshot](https://user-images.githubusercontent.com/38045873/72055321-897df900-32c2-11ea-9613-56b833615c67.jpg)

**Comment by mojavelinux** [comment]
Asciidoctor PDF is centering the image on the line (emulating `vertical-align: middle`). There's currently no way to control the vertical alignment of inline images in Asciidoctor PDF. We could consider adding that as a feature. What you seem to be after is `vertical-align: text-bottom`).

**Comment by mojavelinux** [comment]
> and it gets worse when there are subscripts

To a certain degree, this is just the way those are going to be (once the image can be aligned to text-bottom). Asciidoctor PDF doesn't see text with subscript. It just sees an image. So the bottom of the image, no matter where it is, would align with the base of the text. That's just a limitation of this integration approach.

**Comment by davidplowman** [comment]
Yes, I can see that for each embedded image you'd have to know the vertical coordinate of where the base of the text is supposed to be. Tricky! But thanks for the feedback.

**Comment by mojavelinux** [comment]
One thing I would recommend for asciidoctor-mathematical is to allow the image to be scaled by a fixed amount. That would allow the image to take up more of the line and thus align better with the existing text. I'm not sure how it determines the sizes now, but if they aren't adjustable, that itself can be problematic.

**Comment by ProgramFan** [comment]
I would like to add logic in asciidoctor-mathematical for a better inline equation layout. The feature I would like to have from asciidoctor-pdf, maybe from asciidoctor for inline images in general, is like the following: 
1. allow setting image height to something like: `X line-height`, where x is a floating-point number such as 1.0.
2. allow adjusting the way inline images vertically align with the line, maybe something like 'center, bottom, top'.

This feature can only be provided as an extension to the inline image macro at the moment since inline markups are not part of the AST.

**Comment by ProgramFan** [comment]
In fact, I once tried to make the equations scale with the text, to make them the same size as the texts in titles, but without any successes. So at the moment, the equations in titles look extremely ugly since they are too small.

**Comment by mojavelinux** [comment]
In the current version of Asciidoctor PDF, it's possible to make the image fit to the height of the line using `fit=line`. However, this includes the line spacing, so the image ends up being taller than the text.

I agree we can consider having control over the vertical alignment. Like I said, what you probably want is baseline so that the image lines up with the bottom of the text. (It already does center, and top and bottom would be above and below the text, respectively).

There are, however, limits to how much this can even work. Prawn doesn't really have the same concept of lines like HTML/CSS. So we can fake it to a point, but then it gets really fuzzy.

**Comment by mojavelinux** [comment]
There are actually two problems here. First is the vertical alignment of the image. Second is that there doesn't seem to be any way to influence the height of the inline image that Asciidoctor Mathematical generates. It only bases the size on ppi for PNG and zoom for SVG. That's where the `fit=line` option (or similar) might help...because it allows Asciidoctor PDF to adjust the size of the image.

Perhaps we need `fit=line` and `fit=font-size` depending on whether you want to fit the image to the line height or height of text (capital letter).

I don't think Asciidoctor Mathematical should manipulate the height and width itself (aside from the global scaling). It should be reporting the intrinsic height and width, but also include a role or option that allows the converter (or browser) to make the final adjustment.

**Comment by mojavelinux** [comment]
I'm starting to think a role might be best so that the HTML converter can control the size with CSS.

@ProgramFan Would you also consider adding support for the mathematical-zoom attribute to control the base zoom for SVG images?

**Comment by ProgramFan** [comment]
I can add the mathematical-zoom attribute, but how shall it be defined?

**Comment by mojavelinux** [comment]
I would just allow the user to define `mathematical-zoom`, convert it to a double, and pass it through to Mathematical using the zoom option.

**Comment by ProgramFan** [comment]
I don't quite understand the idea. Here is my confusion: asciidoctor-mathematical works as a tree-preprocessor and just replaces the stem macros with image macros, so it generates the images prior to asciidoctor-pdf. But the zoom value can only be determined in asciidoctor-pdf and can be different under different scenarios (for example in titles and paragraphs). How can asciidoctor-mathematical act on a value that is not defined?

My first image of a solution is that the backends act on things like `height=1em`. And asciidoctor-pdf just set these attributes on the images it generates. But this is just my own opinion.

**Comment by mojavelinux** [comment]
> My first imagnation of a solution is that the backends act on things like height=1em.

Yes, I get that. But for environments for where that doesn't apply or it can't work, the zoom would be a way to influence the intrinsic size that mathematical produces. It's a global setting, defined by the `mathematical-zoom` attribute and passed to the Mathematical constructor.

Regardless, we can't use `height=1em` as that's not a valid value for height. But we could use an attribute like `fit=1em` or a role like `.fit-content` or something.

**Comment by ProgramFan** [comment]
I see. I will add mathematical-zoom just like that.

**Comment by mojavelinux** [comment]
:+1:

We'll then continue this discussion about how to coordinate so that the converter can resize the equation to fit with the size of the text.

**Comment by mcp292** [comment]
Has any progress been made on this front?

**Comment by mojavelinux** [comment]
If there were progress, it would be reported here.

**Comment by mcp292** [comment]
Oh, got it!

**Comment by mojavelinux** [comment]
Related to https://github.com/asciidoctor/asciidoctor-mathematical/issues/45.

**Comment by aburkov** [comment]
Hi! Is there a version of asciidoctor where the inline math characters are aligned with the text in the PDF output?

**Comment by mojavelinux** [comment]
The current state is reflected by this issue and the linked one. If there was a change, this issue would be updated or closed.

If you need help with a workaround, please ask in the project chat at https://chat.asciidoctor.org. Someone there may be able to help you.

**Comment by a4z** [comment]
It would be nice if there could be a solution.

If I make a PDF in slide layout, and change the default fonts, the generated SVG is embedded in the original size, and I have no way to do anything about this, since this seems to happen in asciidoctor-pdf 

<img width="1417" alt="Screenshot 2024-11-03 at 09 04 56" src="https://github.com/user-attachments/assets/5462d975-165d-44cf-96dc-30c57c2212bd">

I had a very similar issue, PDF formula rendering, at a workplace, and this was the issue that I had to stick to another toolchain there.

**Comment by ProgramFan** [comment]
I guess what we really need is a proper engine which can layout images with height be the same as the line height. For block math, the image shall be even larger. asciidoctor-mathematical renders inline and block math as fix-sized svg images, then rely on the backend to adapt the size. asciidoc, as far as I know, has not yet provided a way to enforce the required behavior for images.

**Comment by Crown0815** [comment]
I am also trying to solve this issue mathematically. Is there an information given by Asciidoctor about the ascender and descender height?
Or another way to determine where the bottom of the x-height is?

**Comment by Crown0815** [comment]
I created a gem called [asciidoctor-pdf-mathjax](https://github.com/Crown0815/asciidoctor-pdf-mathjax), which will use MathJax to translate the equations and lay them out properly into the line.
You can find some [examples here](https://github.com/Crown0815/asciidoctor-pdf-mathjax/blob/main/test/verification/stem_latex.verified.pdf).

Font color, baseline alignment, headers, etc. are all taken into account.
I am pretty sure I missed some edge cases, so please let me know if the gem works for you and if you find any issue with it.

**Comment by ProgramFan** [comment]
I think this can be codesigned with asciidoctor-pdf-mathjax, since mathjax is more compable than mathematical's backend. Thanks everyone for the patience and contribution. @Crown0815 @mojavelinux

**Comment by Crown0815** [comment]
> I honestly just don't know how to solve this problem. We could offer something like `fit=x-height`. However, we can't reliably calculate the x-height of the font. So that solution can't even get out of the gate. But even if we could, then it would normalize all images to fit in the x-height, which would mean that a "w" and large "w" would essentially look the same.
> 
> We could just assume that the x height is roughly half the size of the font, I could implement `fit=x-height` and we can move forward with that. The placement isn't going to be pixel perfect because Prawn is just not good enough (it's no LaTeX). But it would produce reasonable behavior.

I believe you can get the x-height from the font:

```ruby
    font = TTFunk::File.open(font_file)
    descender_height = font.horizontal_header.descent.abs
    ascender_height = font.horizontal_header.ascent.abs
    x_height = font.os2.x_height

    unless x_height
      logger.debug "'OS/2' table not found, falling back to estimating font x-height (ex) from glyph"

      cmap_table = font.cmap.tables.find { |table| table.format == 4 && table.platform_id == 3 && table.encoding_id == 1 || table.encoding_id == 10 }
      raise 'No suitable Unicode cmap table found' unless cmap_table

      glyph_id = cmap_table.code_map['x'.ord]
      raise "Glyph for 'x' not found" if glyph_id.nil? || glyph_id == 0

      glyph = font.glyph_outlines.for(glyph_id)
      raise 'Glyph data not available' unless glyph

      x_height = glyph.y_max - glyph.y_min
    end
    logger.debug "Embedding Font: #{font_family} #{font_style}, x-height: #{x_height}, ascender: #{ascender_height}, descender: #{descender_height}"
```

Tested this and got:

- `Embedding Font: Noto Serif bold, x-height: 1098, ascender: 2189, descender: 600`
- `Embedding Font: Roboto-Light bold, x-height: 1082, ascender: 1900, descender: 500`

(in both cases the OS/2 table was available, but I double checked the `unless x_height` algorithm and got the same results for both fonts)

**Comment by mojavelinux** [comment]
I used the word reliably for a reason. Prawn stores the x height on the
front object, but it simply does give us the correct value. And it also
doesn't tell us where the baseline is.

You're welcome to try to prove it. I tried and it just doesn't work.

As I said, I'm willing to implement fit=x-height as half the font height.
Otherwise, I have no solution.

**Comment by Crown0815** [comment]
The baseline you can calculate from the ascender and descender (also in the code example I attached above)

I already did this in asciidictor-pdf-MathJax. You do get pixel-perfect alignment if done right.

The main challenge is if the math is higher than the line height (e.g. with index and power).

Plus, for perfect alignment you need the baseline for the thing you are aligning. Luckily MathJax provides the baseline position.

**Comment by mojavelinux** [comment]
If you're willing, I'd appreciate if you sent a PR, even with a rough
implementation. I couldn't get it to work.

Whether it's mathematical or MathJax, it should make no difference.
Aligning an inline image to the baseline is the same need.
